File: clojure/largest-series-product/src/largest_series_product.clj
(ns largest-series-product)

(defn- char->int
  [c]
  (- (int c) (int \0)))

(defn largest-product [n ds]
  (cond
    (re-find #"[^0-9]" ds)
    (throw (IllegalArgumentException. "Invalid value given"))

    (> n (count ds))
    (throw (IllegalArgumentException. "Span cannot be longer than string"))

    (neg? n)
    (throw (IllegalArgumentException. "Span cannot be negative"))

    :else
    (let [ints (mapv char->int ds)]
      (loop [start 0
             end   n
             max-p 0]
        (if (> end (count ints))
          max-p
          (let [product (apply * (subvec ints start end))]
            (recur (inc start)
                   (inc end)
                   (max max-p product))))))))

File: python/largest-series-product/largest_series_product.py
def largest_product(series, size):
    """
    Given a sequence of digits, looks at sequences of adjacent digits contained
    within the input sequence and returns the largest product from those sequences.
    """
    if len(series) < size:
        raise ValueError("span must be smaller than string length")

    if not series.isdigit():
        raise ValueError("digits input must only contain digits")

    if not size > 0:
        raise ValueError("span must not be negative")

    max_product = 0

    for start in range(0, len(series) - size + 1):
        product = 1
        end = start + size
        for n in list(map(int, series[start:end])):
            product *= n

        max_product = max(product, max_product)

    return max_product
